macaroon_service.go

```diff
@@ -231,12 +234,46 @@ func (ms *MacaroonService) Start() error {
 
 	// There are situations in which we don't want a macaroon to be created
 	// on disk (for example when running inside LiT stateless integrated
-	// mode). For any other cases, we create macaroon files in the default
-	// directory.
-	if ms.cfg.StatelessInit || lnrpc.FileExists(ms.cfg.MacaroonPath) {
+	// mode).
+	if ms.cfg.StatelessInit {
 		return nil
 	}
 
+	// If there are no required permissions, then we don't have to generate
+	// or re-generate a macaroon and so we can exit here.
+	if len(ms.cfg.RequiredPerms) == 0 {
+		return nil
+	}
+
+	// If we are not in stateless mode and a macaroon file does exist, we
+	// check that the macaroon matches the required permissions. If not, we
+	// will delete the macaroon and create a new one.
+	if lnrpc.FileExists(ms.cfg.MacaroonPath) {
+		matches, err := ms.macaroonMatchesPermissions()
+		if err != nil {
+			return fmt.Errorf("unable to match the previous "+
+				"macaroon's permissions with the required "+
+				"permissions: %v", err)
+		}
+
+		// In case the old macaroon matches the required permissions,
+		// we don't need to create a new macaroon.
+		if matches {
+			return nil
+		}
+
+		// Else if the required permissions have been updated, we delete
+		// the old macaroon and create a new one.
+		log.Infof("Macaroon at %s does not have all required "+
+			"permissions. Deleting it and creating a new "+
+			"one", ms.cfg.MacaroonPath)
+
+		if err := os.Remove(ms.cfg.MacaroonPath); err != nil {
+			return fmt.Errorf("Unable to remove macaroon "+
+				"at %s: %v", ms.cfg.MacaroonPath, err)
+		}
+	}
+
 	// We don't offer the ability to rotate macaroon root keys yet, so just
 	// use the default one since the service expects some value to be set.
```

Comment:
also, below on line 301 where we remove the file if writing to it failed: this was ok before because we would only get to that part if the file did not exist in the first place. But now, since we might be overwriting a file that did already exist, we should not remove it in that case if we fail

Reply:
I pushed this as a separate fixup commit right now, as I'd like to get your feedback on it before squashing it:
I see a risk if we remove this file removal code, that the Write code might fail during the execution for various reasons, corrupting the macaroon file. If that would happen, we'd now end up in a state where the permission comparison code will fail to parse it to a macaroon, and we'd never write a new macaroon file without manual removal of the file?

Reply:
i think that is ok - ie to force a manual removal of the file in these rare cases. I'd rather the user make this decision than assume that we can remove it

Reply:
Ok sounds good :)! Squashing the commit then.

Reply:
Squashed with the latest push.